### PR TITLE
[TVMC] Allow manual shape specification in tvmc

### PR DESCRIPTION
**Pull request by CircleSpin**
Currently tvmc does not allow users to overriding specific shapes (such as batch sizes) and instead uses what is defined within the model file. In most cases this is most practical, but in there are some cases that would benefit from additional flexibility i.e. undefined shape sizes, fiddling with batch sizes. This PR adds support for manually specifying input shapes. :)

### CODE REVIEW

**Comment by CircleSpin** [comment]
@leandron @jwfromm @mbrookhart 

Could you take a look at this PR and let me know what you think?

**Review by leandron — COMMENTED** [review]
Hi @CircleSpin, thanks for the PR - I think shape definition from the outside is something we want to have in `tvmc`. So much so, that we have discussions going on in #7359.

I think we can use some ideas from the ongoing discussion in #7359, like the syntax to define shapes, and apply shape definition to more frontends, as you have covered here.

cc @ekalda @u99127

**Comment by comaniac** [comment]
Agree with @leandron. We can actually just keep one PR for both purposes, because PyTorch fix is more like a corner case of supporting custom shape.

**Comment by jwfromm** [comment]
Funny that these two very similar PRs landed so close to one another. After reading both, I would argue in favor of choosing this implementation. I think its important to be able to specify a mapping between input name and shape. Not only is this important for constructing a shape dictionary for all non-pytorch frontends, but it also allows a subset of shapes to be overwritten. If a bert model for example had 4 default shapes, this approach would allow us to specify `seq_len=[256]` while leaving the other 3 shapes alone.

**Comment by comaniac** [comment]
@CircleSpin @ekalda could you folks organize a plan to work on the one PR and close the other?

**Comment by ekalda** [comment]
From what I can tell, the functionality this PR implements is a superset of the one in my PR, so I'm happy to close mine.

File: python/tvm/driver/tvmc/compiler.py [hunk 1]
```diff
@@ -36,6 +36,41 @@
...
+        for x in string[1]:
+            x = int(x)
+            # Negative numbers are converted to dynamic axes.
+            if x < 0:
+                x = relay.Any()
+            shapelist.append(x)
+        # Assign dictionary key value pair.
+        d[string[0]] = shapelist
+    return d
+
```
Comment (ekalda) [inline python/tvm/driver/tvmc/compiler.py:72]: Maybe it would be good to add some error handling when the user input is not in the desired format? Also, should we maybe allow inputs that are "close", e.g. "name:num1Xnum2" or "name:num1xnum2, name2:num1xnum2" ? Some unit tests to handle the corner cases would be nice :)

File: python/tvm/driver/tvmc/frontends.py [hunk 2]
```diff
@@ -297,6 +303,15 @@ def load(self, path):
         traced_model.eval()  # Switch to inference mode
         input_shapes = [("input{}".format(idx), shape) for idx, shape in enumerate(shapes)]
```
Comment (ekalda) [inline python/tvm/driver/tvmc/frontends.py:304]: (1) That "shapes" in "enumerate(shapes)" is not defined (and old bug, something I discovered when I tried to run a PyTorch model with tvmc).
(2) This approach of extracting input shapes form PyTorch model is not functional any more (see the discussion in #7359 ), so this parameter needs to be set for PyTorch and the PyTorch frontend should throw an error when it is not set. (Unless someone maybe knows a way to extract the shapes from the model?)

File: tests/python/driver/tvmc/test_compiler.py [hunk 3]
```diff
@@ -56,6 +53,15 @@ def test_compile_tflite_module(tflite_mobilenet_v1_1_quant):
     assert type(dumps) is dict
 
 
+def test_compile_tflite_module(tflite_mobilenet_v1_1_quant):
+    # Check default compilation.
+    verify_compile_tflite_module(tflite_mobilenet_v1_1_quant)
+    # Check with manual shape override
+    shape_string = "input:1x224x224x3"
+    shape_dict = tvmc.compiler.parse_shape(shape_string)
+    verify_compile_onnx_module(tflite_mobilenet_v1_1_quant, shape_dict)
```
Comment (ekalda) [inline tests/python/driver/tvmc/test_compiler.py:62]: Should this call verify_compile_tflite_module?

**Review by ekalda — CHANGES_REQUESTED** [review]
Some suggestions (mostly around PyTorch), but I agree with others that it is good to have --shapes as a general parameter :)

Another suggestion (I didn't know where was a good place to add it):
I think it would make sense to add --shapes parameter to tvmc tune as well since tvmc tune also calls frontend.load_model. In fact, since PyTorch models rely on that parameter having a value, we wouldn't be able to tune them without that parameter being part of tvmc tune.

**Comment by CircleSpin** [comment]
Thank you @ekalda for your super thorough feedback! :trophy: We've tried to incorporate it, and also included a test in test_frontends.py that imports a torchvision resnet18 using the new shape specification. I've also added shape inputs to tuning as you had mentioned. Could you take another look and let me know if you have any additional thoughts or feedback? :smile:

File: python/tvm/driver/tvmc/common.py [hunk 4]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+
+    Returns
+    -------
+    shape_dict: dict
+        A dictionary mapping input names to their shape for use in relay frontend converters.
+    """
+    inputs = inputs.replace(" ", "")
+    # Check if the passed input is in the proper format.
+    valid_pattern = re.compile("(\w+:(\d+(x|X))*(\d)+)(,(\w+:(\d+(x|X))*(\d)+))*")
+    result = re.fullmatch(valid_pattern, inputs)
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:160]: - Seems doesn't match negative numbers such as `data:-1x3x224x224`.
- Same comment as the one I left in another PR: This syntax is not straightforward to many users. I would suggest using either JSON or Python syntax.

File: python/tvm/driver/tvmc/common.py [hunk 5]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+    if result is None:
+        raise argparse.ArgumentTypeError(
+            "--shapes argument must be of the form 'input_name:dim1xdim2x...xdimN,input_name2:dim1xdim2"
+        )
+    d = {}
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:165]: Please avoid bad naming.

File: python/tvm/driver/tvmc/common.py [hunk 6]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+    # Break apart each specific input string
+    inputs = inputs.split(",")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:167]: Do not mess up with types. `inputs` is already a string.

File: python/tvm/driver/tvmc/common.py [hunk 7]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+    for string in inputs:
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:168]: Please avoid bad naming. `string` is too general and looks like a preserved word (although it's not).

File: python/tvm/driver/tvmc/common.py [hunk 8]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+        # Split name from shape string.
+        string = string.split(":")
+        shapelist = []
+        # Separate each dimension in the shape.
+        string[1] = string[1].lower().split("x")
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:173]: Ditto. Do not mess up with types.

File: python/tvm/driver/tvmc/common.py [hunk 9]
```diff
@@ -136,3 +138,46 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
+        # Parse each dimension into an integer.
+        for x in string[1]:
+            x = int(x)
+            # Negative numbers are converted to dynamic axes.
+            if x < 0:
+                x = relay.Any()
+            shapelist.append(x)
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:180]:
```suggestion
            shapelist.append(x if x >= 0 else relay.Any())
```

File: python/tvm/driver/tvmc/compiler.py [hunk 10]
```diff
@@ -158,6 +167,9 @@ def compile_model(
         The layout to convert the graph to. Note, the convert layout
         pass doesn't currently guarantee the whole of the graph will
         be converted to the chosen layout.
+    shape_dict: dict, optional
+        A mapping between input names and their shape. This is useful
+        to override the default values in a model if needed.
```
Comment (comaniac) [inline python/tvm/driver/tvmc/compiler.py:172]:
```suggestion
        A mapping from input names to their shapes. When present,
        the default shapes in the model will be overwritten.
```

File: python/tvm/driver/tvmc/frontends.py [hunk 11]
```diff
@@ -285,17 +291,18 @@ def suffixes():
...
         # pylint: disable=C0415
         import torch
 
         traced_model = torch.jit.load(path)
+        traced_model.eval()  # Switch to inference mode
 
-        inputs = list(traced_model.graph.inputs())[1:]
-        input_shapes = [inp.type().sizes() for inp in inputs]
+        if shape_dict is None:
+            raise TVMCException("--shapes must be specified for {}".format(self.name()))
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:302]:
```suggestion
            raise TVMCException("--shapes must be specified for %s" % self.name())
```

File: python/tvm/driver/tvmc/frontends.py [hunk 12]
```diff
@@ -54,13 +54,15 @@ def suffixes():
...
-    def load(self, path):
+    def load(self, path, shape_dict=None):
         """Load a model from a given path.
 
         Parameters
         ----------
         path: str
             Path to a file
+        shape_dict: dict, optional
+            A dictionary mapping input names to shapes.
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:65]:
```suggestion
            Mapping from input names to their shapes.
```

File: python/tvm/driver/tvmc/frontends.py [hunk 13]
```diff
@@ -389,6 +396,8 @@ def load_model(path, model_format=None):
     model_format : str, optional
         The underlying framework used to create the model.
         If not specified, this will be inferred from the file type.
+    shape_dict : dict, optional
+        A mapping between input names and their desired shape.
```
Comment (comaniac) [inline python/tvm/driver/tvmc/frontends.py:400]:
```suggestion
        Mapping from input names to their shapes.
```

File: tests/python/driver/tvmc/test_common.py [hunk 14]
```diff
@@ -149,3 +149,27 @@ def test_tracker_host_port_from_cli__only_hostname__default_port_is_9090():
 
     assert expected_host == actual_host
     assert expected_port == actual_port
+
+
+def test_shape_parser():
```
Comment (comaniac) [inline tests/python/driver/tvmc/test_common.py:154]: Cover negative shapes.

**Review by comaniac — CHANGES_REQUESTED** [review]

**Comment by comaniac** [comment]
Also cc @masahi to review PyTorch shape handling.

File: tests/python/driver/tvmc/test_common.py [hunk 15]
```diff
@@ -149,3 +149,27 @@ def test_tracker_host_port_from_cli__only_hostname__default_port_is_9090():
...
+    assert shape_dict == {"input": [10, 10, 10], "input2": [20, 20, 20, 20]}
+
+    # Check that invalid pattern raises expected error.
+    shape_string = "input:ax10"
+    with pytest.raises(argparse.ArgumentTypeError):
+        tvmc.common.parse_shape_string(shape_string)
+    # Check that input with invalid separators raises error.
+    shape_string = "input:5,10 input2:10,10"
+    with pytest.raises(argparse.ArgumentTypeError):
+        tvmc.common.parse_shape_string(shape_string)
```
Comment (leandron) [inline tests/python/driver/tvmc/test_common.py:175]: I suggest splitting these test cases above in specific unit test - I see 5 self-contained unit tests here.

File: tests/python/driver/tvmc/test_compiler.py [hunk 16]
```diff
@@ -56,6 +53,15 @@ def test_compile_tflite_module(tflite_mobilenet_v1_1_quant):
     assert type(dumps) is dict
 
 
+def test_compile_tflite_module(tflite_mobilenet_v1_1_quant):
+    # Check default compilation.
```
Comment (leandron) [inline tests/python/driver/tvmc/test_compiler.py:57]:
```suggestion
def test_compile_tflite_module(tflite_mobilenet_v1_1_quant):
    # some CI environments wont offer flute, so skip in case it is not present
     pytest.importorskip("tflite")
```

Not all CI environments will offer `tflite`.

File: python/tvm/driver/tvmc/autotuner.py [hunk 17]
```diff
@@ -210,6 +210,13 @@ def add_tune_parser(subparsers):
     #     can be improved in future to add integration with a modelzoo
     #     or URL, for example.
     parser.add_argument("FILE", help="path to the input model file")
+    parser.add_argument(
+        "--shapes",
```
Comment (leandron) [inline python/tvm/driver/tvmc/autotuner.py:214]: Question: is the terminology `--shapes` specific enough for people to always infer they are "input shapes"? I'm asking because in the PR we previously had in discussion, we called it `--input-shapes`, and just want to hear what others think.

cc @comaniac @ekalda

**Review by leandron — CHANGES_REQUESTED** [review]
Some comments, trying not to overlap with other reviews.

File: python/tvm/driver/tvmc/autotuner.py (hunk 17 above)
Comment (comaniac) [reply python/tvm/driver/tvmc/autotuner.py:214]: input-shapes is better.

**Comment by CircleSpin** [comment]
@hogepodge

File: python/tvm/driver/tvmc/common.py (hunk 4 above)
Comment (jwfromm) [reply python/tvm/driver/tvmc/common.py:160]: @comaniac I agree the syntax is a little confusing, but discussed it with Jocelyn and there are some tradeoffs that would be made using a more Python/JSON like syntax. Since we're passing the shapes argument via command line, we cant use spaces or semicolons unless the whole argument is wrapped in quotes. For example, if we used a python like syntax such as `tvmc compile --shapes=data:-1,3,224,224 data2:10,10` the space would prevent argparse from recognizing it as a single argument. We could require quote wrapping but I personally think that is more likely to cause issues than using `x` instead of `,` to separate dimensions. Do you have a recommendation for the input format that you prefer or are you ok with always wrapping in quotes?

Comment (comaniac) [reply python/tvm/driver/tvmc/common.py:160]: To be honest, I'm totally fine with quotes in the command line, and we actually use it in TVMC already. For example, when specifying target, you cannot avoid using quotes for targets like `llvm -mcpu=core-axv2`. In fact, although we haven't fully landed the new target, the ultimate goal is to get rid of plain string target format and support JSON format. IMHO, it would be better if the format/syntax are consistent.

Meanwhile, if having quotes in the command line is a concern, it might be worthwhile to consider taking a command line file. See https://docs.python.org/3/library/argparse.html#fromfile-prefix-chars for reference. In this case, we can have:

```
# shapes.txt
--input-shapes
data:[-1,3,224,224] data2: [10,10]
```

```
tvmc compile @shapes.txt
```

Comment (leandron) [reply python/tvm/driver/tvmc/common.py:160]: > See https://docs.python.org/3/library/argparse.html#fromfile-prefix-chars for reference

That's really cool @comaniac I didn't know this.

I agree that `data:[-1,3,224,224] data2: [10,10]` syntax looks clearer, IMHO.

File: tests/python/driver/tvmc/test_common.py (hunk 15 above)
Comment (jwfromm) [reply tests/python/driver/tvmc/test_common.py:175]: I actually really prefer having these grouped under a single test since they're all testing functionality of the same simple function and breaking them into separate tests would add a lot of bloat for little benefit. @comaniac, which way do you prefer?

Comment (comaniac) [reply tests/python/driver/tvmc/test_common.py]: Since all of them are testing `parse_shape_string`, I prefer to put them together as it currently does. It seems not a big deal to locate the problematic corner case within this unit test when it fails.

**Comment by CircleSpin** [comment]
@comaniac @leandron @jwfromm 

The code has been refactored. The input syntax has been updated so that "data:[32,3,224,224] data2:[1,4,4,4]" will work, and the previous one no longer will. Could you take another pass and let me know your thoughts on this new version?

File: python/tvm/driver/tvmc/common.py [hunk 18]
```diff
@@ -136,3 +138,40 @@ def tracker_host_port_from_cli(rpc_tracker_str):
...
     return rpc_hostname, rpc_port
+
+
+def parse_shape_string(inputs_string):
+    """Parse an input shape dictionary string to a usable dictionary.
+
+    Parameters
+    ----------
+    inputs_string: str
+        A string of the form "name:num1xnum2x...xnumN,name2:num1xnum2xnum3" that indicates
```
Comment (comaniac) [inline python/tvm/driver/tvmc/common.py:149]: Update the syntax.

**Review by comaniac — APPROVED** [review]
LGTM. Just a nit.

File: python/tvm/driver/tvmc/frontends.py [hunk 19]
```diff
@@ -285,17 +291,18 @@ def suffixes():
...
         # pylint: disable=C0415
         import torch
 
         traced_model = torch.jit.load(path)
+        traced_model.eval()  # Switch to inference mode
 
-        inputs = list(traced_model.graph.inputs())[1:]
-        input_shapes = [inp.type().sizes() for inp in inputs]
+        if shape_dict is None:
+            raise TVMCException("--input-shapes must be specified for %s" % self.name())
```
Comment (leandron) [inline python/tvm/driver/tvmc/frontends.py:302]: A suggestion here would be moving this validation/error msg to be checked before the model is loaded (line  297 maybe?), just so that the user don't wait the time for a (sometimes big) model to be loaded, just to be presented with an error message that we could've checked already.

**Review by leandron — COMMENTED** [review]
LGTM, with just a small comment.

**Comment by CircleSpin** [comment]
Thanks for the catches you both found, @comaniac and @leandron :smiley:

**Review by jwfromm — APPROVED** [review]
This LGTM. Thanks for making this change @CircleSpin, it will definitely be helpful to TVMC users!

**Review by leandron — APPROVED** [review]
LGTM - thanks for the effort in implementing this feature @CircleSpin.

**Comment by leandron** [comment]
I think this one is ready for merge, right?

**Comment by comaniac** [comment]
> I think this one is ready for merge, right? Just need another CI run...

@CircleSpin please fix or re-trigger the CI.

**Comment by jwfromm** [comment]
@masahi the previous build error seems like a torch failure unrelated to TVM but is not at all descriptive. Have you run into that one before?

**Comment by comaniac** [comment]
> @masahi the previous build error seems like a torch failure unrelated to TVM but is not at all descriptive. Have you run into that one before?

It's not related to torch failure. It was because of the 2-hour timeout and has been worked around.

**Comment by jwfromm** [comment]
Looks like the new commit failed again in the same way, python aborting during construction of the torchvision model.

**Comment by masahi** [comment]
Yeah looks wierd. Does it reproduce locally? Maybe CPU node has older torch and torchvision

**Comment by jwfromm** [comment]
this works fine locally with the same version of torch and torchvision as CI. Not sure what the deal is.

**Comment by masahi** [comment]
I don't think doing frontend tests during CPU unittest is a good idea. I suggest moving `test_frontends.py` to `test/python/frontend`.

**Comment by leandron** [comment]
> I don't think doing frontend tests during CPU unittest is a good idea. I suggest moving `test_frontends.py` to `test/python/frontend`.

Some test needs to be done to exercise the integration between TVMC and the underlying APIs - we want to keep this to a minimum, and will be reducing the number of tests that actually do this as per previous discussions in #6831 (cc @d-smirnov).

In #7304 I have the first test within TVMC that mocks all the APIs and just checks for arguments being passed - I intend to use that one as an example to motivate some refactoring in the TVMC tests to make them easier/faster to run.

**Comment by leandron** [comment]
@comaniac it looks like now the CI issue are gone, can we merge this one when you have a moment?

**Comment by comaniac** [comment]
Thanks @CircleSpin @ekalda @leandron @jwfromm @masahi

**Comment by apivovarov** [comment]
Quite ofter I see the following failure in CI builds
```
test_frontends.py::test_load_model__pth munmap_chunk(): invalid pointer
Fatal Python error: Aborted
```
`test_load_model__pth` was added in this PR

https://github.com/apache/tvm/issues/7471
